### ISSUE TITLE
Links to data privacy statements

### ISSUE THREAD

**Issue by tumbl3w33d** [kanidm/kanidm#2171]
**Is your feature request related to a problem? Please describe.**

Users need to be informed how you work with their data (GDPR etc.). Kanidm cannot provide a default, because it depends on hosting provider, backup strategy, settings, etc., however, kanidm can support it by providing links in the right place(s).

**Describe the solution you'd like**

An admin-configurable URL (app config file would do, no need to configure via web UI). It would show up when users start using kanidm, like in the onboarding via reset link. Maybe there would be the option to enforce that the user confirms they agree/have seen/read it.

In addition, but that's a bonus in my view, I would like to configure such URL per oauth2 app, so it would show up when the user consents with which data will be shared (i.e., oauth2 roundtrip). This would be stored in the DB with the oauth2 config and set via API.

**Describe alternatives you've considered**

I could rewrite the markup in a proxy to hack my link in, but eww...

**Comment by yaleman** [comment]
> configure such URL per oauth2 app, so it would show up when the user consents with which data will be shared (i.e., oauth2 roundtrip). This would be stored in the DB with the oauth2 config and set via API.

This feels like features for features' sake - is this typical in Europe/other places? I work for a global company and if I was prompted to agree to using other services (from a privacy perspective, not authorisation) when trying to log in I'd assume I was being phished. Worse still, if I disagreed, I'd be opting out of a job... 😄 

Your organisation agrees to using the third-party system, and it's up to the organisation to negotiate those terms of service on behalf of the user, and the user to consent to working in the organisation's environment. The RS should have its own onboarding flow for this in either case.

**Comment by tumbl3w33d** [comment]
Yeah, that's quite European (EU + friends). It's similar to the cookie consent crap on every website nowadays. It's all about covering your ass as an enterprise providing a service that works with personal data.

I'm also quite sure that a company you work for cannot waive your right to informational self-determination, because you are not a slave. But law is not my profession, so I'm not gonna go down that rabbit hole other than making sure I reduce the legal attack surface.

As I said, the Oauth2 part would just be convenient, because it would be a hook to keep the user out if they really disagree with their data being processed. You are right that I can integrate it into every connected app, which is not always easy because - like you - most people aren't Europeans and are completely confused by these crazy things we come up with.

Nobody reads all these things and most people don't care about their data privacy. It's all about having it written down and showing to the user so you're not getting sued. They call it being GDPR-compliant and you can write that as USP on your product. ;)

**Comment by Firstyear** [comment]
We already have a hook/place for this surprisingly, in the oauth2 scope consent grant. The first time you follow an oauth2 link you are clearly notified what information the target site can access and you are given a chance to consent to proceed or not. In addition, if those scopes *change* you are prompted to consent again.

So even though we don't explicitly call out GDPR on the page, we are clearly giving the user consent and determination over how their data is used and transmitted to the oauth2 target application. 

From my view, about the only other thing we could do is have a way on the consent page to link to a deployment specific policy, but even then I'm not sure about that because the privacy policy needs to be on the third party (oauth2 target site) not kanidm. 

From my view, I actually think we don't need to do anything here because really what you need is a privacy statement on the *kani* site itself, and then a seperate one on the oauth2 application. Since these are the two points that are then handling the information in question. 

Finally I think rather than just commit code to "maybe solve a legal problem" the best thing to do is actually consult with a legal team about what *really* needs to happen here.

**Comment by yaleman** [comment]
I'd agree an optional link for a system-wide data privacy statement could be handy though - probably on the login page down next to the "powered by Kanidm" footer?

**Comment by Firstyear** [comment]
Would it be a link? Or text an admin would upload?

**Comment by yaleman** [comment]
Most orgs *should* have a web site (/ shared google doc)  with a central one to refer to... 🤷🏻

**Comment by Firstyear** [comment]
What other considerations should we have here? I can see that admins may want to have other links and references to things. I'm wondering if there is a generic thing we could do or if we could look at other systems to work out prior art and what requirements might exist?

**Comment by yaleman** [comment]
I'd definitely be looking for prior art on this one - TBH I really don't want to see us ending up hosting editable webpages in the IDP, it's just too huge an attack surface. 🤨

**Comment by Firstyear** [comment]
Agreed. Part of me wonders if we don't even do this, and we just defer it to "make an oauth2 application called policy-centre" that has all the relevant policies for your org.

**Comment by Firstyear** [comment]
@yaleman I think allowing the domain admin to set a single url which links to a privacy policy would suffice here. Then we can display it in the footer near the "powered by kanidm". Does that seem reasonable?

**Comment by yaleman** [comment]
Seems legit

**Comment by tacerus** [comment]
Maybe custom footer links could be made more flexible and not be constrained to a single "Privacy policy" option? I could see adding other links (impress, ticket system) being useful as well.
Not sure how many customization toggles there should be before allowing users to set their own page templates though. :)
